man/chfn.1
.TH CHFN 1 "1 Feb 1988" "Project Athena"
\" RCSID: $Header: /afs/athena.mit.edu/astaff/project/moiradev/repository/moira/man/chfn.1,v 1.7 1998-05-26 18:14:10 danw Exp $
.SH NAME
chfn \- change finger information
.SH SYNOPSIS
.B chfn
[ username ]
.SH DESCRIPTION
This command changes (or installs)
the GECOS information field
associated with
.IR username
(your own name by default).  It also asks for some information not in
the standard GECOS field, such as a nickname and home address.
The GECOS information is the information shown by
.I finger(1).
.PP
When invoked,
.B chfn
displays the current information as default values
while prompting for any changes.
The fields include a user's
``real life'' name, nickname, home address and phone number,
office address and phone number, department and year.
Included in each prompt is a default value,
which is enclosed between brackets.
The default value is accepted simply by typing a carriage
return.  To enter a blank field, the word ``none'' may
be typed.  Below is a sample run:
.PP
Info last changed on 22-mar-1988 15:13:33 by user studley using chfn
.br
Name [Biff Studsworth II]:
.br
Nickname [Studley]:
.br
Home address (Ex: EC Bemis 514) [New House 4-601]:
.br
Home phone number (Ex: 3141592) [225-1234]:
.br
Office address (Exs: E40-342 or 2-108) [E40-300M]: \fBE40-358M\fP
.br
Office phone (Ex: 3-7619) []: \fB1863\fP
.br
MIT department (Exs: 9, Biology, Information Services) [VI-3]: \fBnone\fP
.br
MIT year (Exs: 1989, '91, Faculty, Grad) ['90]: \fB\'91\fP
.PP
.I chfn
allows phone numbers to be entered with or without hyphens.
.PP
Note that the changes will not show up with
.I finger
until the day after the database has been changed.
.PP
Users may only change their own information unless they are
Moira database administrators.
.SH AUTHOR
Emanuel 'Jay' Berkenbilt
.SH "SEE ALSO"
moira(1), finger(1)


debian/changelog
debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


clients/chfn/chfn.c
/* $Id: chfn.c,v 1.3 2006-08-23 20:33:33 zacheiss Exp $
 *
 * Talk to the Moira database to change a person's GECOS information.
 *
 * chfn with no modifiers changes the information of the user who is
 * running the program.
 *
 * If a commandline argument is given, it is taken to be the username
 * of the user whose information is to be changed.
 *
 * Copyright (C) 1988-1998 by the Massachusetts Institute of Technology
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 */

#include <mit-copyright.h>
#include <moira.h>
#include <moira_site.h>
#include <mrclient.h>

#include <ctype.h>
#include <stdio.h>
#include <string.h>

RCSID("$Header: /afs/athena.mit.edu/astaff/project/moiradev/repository/moira/clients/chfn/chfn.c,v 1.3 2006-08-23 20:33:33 zacheiss Exp $");

#define FALSE 0
#define TRUE 1

char *whoami;
char *username;

struct finger_info {
  char *fullname;
  char *nickname;
  char *home_address;
  char *home_phone;
  char *office_address;
  char *office_phone;
  char *mit_department;
  char *mit_year;
};

#define argis(a, b) (!strcmp(*arg + 1, a) || !strcmp(*arg + 1, b))

void usage(void);
int get_user_info(int argc, char *argv[], void *message);
char *ask(char *question, char *def_val, int phone_num);
void get_new_info(struct finger_info *old_info, struct finger_info *new_info);

int main(int argc, char *argv[])
{
  int status;			/* general purpose exit status */
  int q_argc;			/* argc for mr_query */
  char *q_argv[F_END];		/* argv for mr_query */
  int i;
  struct finger_info old_info;
  struct finger_info new_info;
  char **arg = argv;
  char *server = NULL;

  if ((whoami = strrchr(argv[0], '/')) == NULL)
    whoami = argv[0];
  else
    whoami++;

  /* parse our command line options */
  while (++arg - argv < argc)
    {
      if (**arg == '-')
	{
	  if (argis("db", "database"))
	    {
	      if (arg - argv < argc - 1)
                {
                  ++arg;
                  server = *arg;
                }
              else
                usage();
	    }
	}
      else if (username == NULL)
	username = *arg;
      else
	usage();
    }

  if (!username)
    {
      username = mrcl_krb_user();
      if (!username)
	exit(1);
    }

  if (mrcl_connect(server, "chsh", 2, 1) != MRCL_SUCCESS)
    exit(1);
  
  /* First, do an access check. */
  
  q_argv[F_NAME] = username;
  for (i = F_NAME + 1; i < F_MODTIME; i++)
    q_argv[i] = "junk";
  q_argc = F_MODTIME;		/* one more than the last updatable field */

  if ((status = mr_access("update_finger_by_login", q_argc, q_argv)))
    {
      com_err(whoami, status, "; finger\ninformation not changed.");
      exit(2);
    }

  printf("Changing finger information for %s.\n", username);

  /* Get information */

  q_argv[NAME] = username;
  q_argc = NAME + 1;
  if ((status = mr_query("get_finger_by_login", q_argc, q_argv,
			 get_user_info, &old_info)))
    {
      com_err(whoami, status, "while getting user information.");
      exit(2);
    }

  /* Get the new information from the user */

  get_new_info(&old_info, &new_info);

  /* Do the update */

  printf("Changing finger information...\n");

  q_argv[F_NAME] = username;
  q_argv[F_FULLNAME] = new_info.fullname;
  q_argv[F_NICKNAME] = new_info.nickname;
  q_argv[F_HOME_ADDR] = new_info.home_address;
  q_argv[F_HOME_PHONE] = new_info.home_phone;
  q_argv[F_OFFICE_ADDR] = new_info.office_address;
  q_argv[F_OFFICE_PHONE] = new_info.office_phone;
  q_argv[F_MIT_DEPT] = new_info.mit_department;
  q_argv[F_MIT_AFFIL] = new_info.mit_year;
  q_argc = F_MODTIME;		/* First non-update query argument */

  if ((status = mr_query("update_finger_by_login", q_argc, q_argv,
			 NULL, NULL)))
    {
      com_err(whoami, status, "while updating finger information.");
      exit(1);
    }

  printf("Finger information updated succesfully.\n");

  mr_disconnect();

  return 0;
}

int get_user_info(int argc, char *argv[], void *message)
{
  struct finger_info *old_info = message;

  if (argc != F_END)
    {
      fprintf(stderr, "Some internal error occurred; try again.\n");
      exit(3);
    }

  printf("Info last changed on %s by user %s with %s.\n",
	 argv[F_MODTIME], argv[F_MODBY], argv[F_MODWITH]);

  old_info->fullname = strdup(argv[F_FULLNAME]);
  old_info->nickname = strdup(argv[F_NICKNAME]);
  old_info->home_address = strdup(argv[F_HOME_ADDR]);
  old_info->home_phone = strdup(argv[F_HOME_PHONE]);
  old_info->office_address = strdup(argv[F_OFFICE_ADDR]);
  old_info->office_phone = strdup(argv[F_OFFICE_PHONE]);
  old_info->mit_department = strdup(argv[F_MIT_DEPT]);
  old_info->mit_year = strdup(argv[F_MIT_AFFIL]);

  /* Only pay attention to the first match since login names are
     unique in the database. */
  return MR_ABORT;
}

char *ask(char *question, char *def_val, int phone_num)
{
  static char buf[BUFSIZ];
  int ok = FALSE;
  char *result;
  int i;
  int dashes = FALSE;
  int len;

#define BLANK "none"

  while (!ok)
    {
      ok = TRUE;
      printf("%s [%s]: ", question, def_val);
      if (!fgets(buf, sizeof(buf), stdin))
	exit(0);
      buf[strlen(buf) - 1] = '\0';
      if (strlen(buf) == 0)
	result = def_val;
      else if (!strcasecmp(buf, BLANK))
	result = "";
      else
	result = buf;

      len = strlen(buf);
      for (i = 0; i < len; i++)
	{
	  switch (buf[i])
	    {
	    case '"':
	      printf("'\"' is not allowed.\n");
	      ok = FALSE;
	      break;
	    case ',':
	      printf("',' is not allowed.\n");
	      ok = FALSE;
	      break;
	    case ':':
	      printf("':' is not allowed.\n");
	      ok = FALSE;
	      break;
	    default:
	      if (iscntrl(buf[i]))
		{
		  printf("Control characters are not allowed.\n");
		  ok = FALSE;
		  break;
		}
	    }
	  if (!ok)
	    break;
	}
    }
  return result;
}

void get_new_info(struct finger_info *old_info, struct finger_info *new_info)
{
  printf("Default values are printed inside of '[]'.\n");
  printf("To accept the default, type <return>.\n");
  printf("To have a blank entry, type the word '%s'.\n\n", BLANK);

#define GETINFO(m, v, n) new_info->v = strdup(ask(m, old_info->v, n))

  GETINFO("Full name", fullname, FALSE);
  GETINFO("Nickname", nickname, FALSE);
  GETINFO("Home address (Ex: EC Bemis 514)", home_address, FALSE);
  GETINFO("Home phone number (Ex: 3141592)", home_phone, TRUE);
  GETINFO("Office address (Exs: E40-342 or 2-108)",
	  office_address, FALSE);
  GETINFO("Office phone (Ex: 3-7619)", office_phone, TRUE);
  GETINFO("MIT department (Exs: 9, Biology, Information Services)",
	  mit_department, FALSE);
  GETINFO("MIT year (Exs: 1989, '91, Faculty, Grad)", mit_year, FALSE);
}

void usage(void)
{
  fprintf(stderr, "Usage: %s [user]\n", whoami);
  exit(1);
}
